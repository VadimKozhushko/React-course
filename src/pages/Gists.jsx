import React from 'react'
import { useDispatch, useSelector } from 'react-redux';

import CircularProgress from '@mui/material/CircularProgress';
import Button from '@mui/material/Button';
import Box from '@mui/material/Box';

import { selectGists, selectGistsError, selectGistsLoading } from '../store/gists/selectors';
import { getAllGists } from '../store/gists/actions';

export const GistsList = () => {

  const gists = useSelector(selectGists);
  const error = useSelector(selectGistsError);
  const loading = useSelector(selectGistsLoading);

  const dispatch = useDispatch();

  const requestGists = () => {
    dispatch(getAllGists());
  };

  if (error) {
    return (
      <>
        <h3>Error</h3>
        <button onClick={requestGists}>Retry</button>
      </>
    );
  }

  return (
    <>
      <h1>GistsList</h1>
      <Button
        variant="contained"
        color="success"
        onClick={requestGists}
      >
        Get API
      </Button>
      {loading && (
        <Box sx={{ display: 'flex', justifyContent: 'center' }}>
          <CircularProgress />
        </Box>
      )}
      {!loading && gists.map((gist) => (
        <Box sx={{ display: 'flex', justifyContent: 'center' }}>
          <li key={gist.id}>{gist.description}</li>
        </Box>
      ))}
      {error && <p style={{ color: 'red' }}>{error}</p>}
    </>
  )
}