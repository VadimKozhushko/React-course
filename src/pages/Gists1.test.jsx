import React from "react";
import { render, screen } from "@testing-library/react";
import '@testing-library/jest-dom'
import userEvent from "@testing-library/user-event";

import { GistsList } from "./Gists";
import * as reduxHooks from 'react-redux'

jest.mock('react-redux')

describe('testing Gist Page', () => {
  const mockUseSelector = jest.spyOn(reduxHooks, 'useSelector')
  const mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch')
  const dispatch = jest.fn()

  it('creating snapshot Gist Pages', () => {
    mockUseSelector.mockReturnValue([])
    mockUseDispatch.mockReturnValue(dispatch)
    const component = {}
      render(
            <GistsList/>     
          )
    expect(component).toMatchSnapshot()
  })
  it('testing button functioan', async () => {
    render(
      <GistsList/>
    )
    const button = screen.getByRole('button')
    await userEvent.click(button)
    expect(dispatch).toHaveBeenCalled()
  })
})