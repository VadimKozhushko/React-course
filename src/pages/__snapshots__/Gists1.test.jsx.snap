// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Gist Page creating snapshot Gist Pages 1`] = `Object {}`;
